src/hotspot/share/prims/scopedMemoryAccess.cpp

```diff
@@ -104,38 +146,62 @@ class CloseScopedMemoryClosure : public HandshakeClosure {
       return;
     }
 
-    frame last_frame = jt->last_frame();
-    RegisterMap register_map(jt,
-                             RegisterMap::UpdateMap::include,
-                             RegisterMap::ProcessFrames::include,
-                             RegisterMap::WalkContinuation::skip);
-
-    if (last_frame.is_safepoint_blob_frame()) {
-      last_frame = last_frame.sender(&register_map);
-    }
-
-    ResourceMark rm;
-    if (last_frame.is_compiled_frame() && last_frame.can_be_deoptimized()) {
-      // FIXME: we would like to conditionally deoptimize only if the corresponding
-      // _session is reachable from the frame, but reachabilityFence doesn't currently
-      // work the way it should. Therefore we deopt unconditionally for now.
-      Deoptimization::deoptimize(jt, last_frame);
-    }
-
     if (jt->has_async_exception_condition()) {
       // Target thread just about to throw an async exception using async handshakes,
       // we will then unwind out from the scoped memory access.
       return;
     }
 
-    if (is_in_scoped_access(jt, JNIHandles::resolve(_session))) {
+    bool in_scoped = false;
+    if (is_accessing_session(jt, JNIHandles::resolve(_session), in_scoped)) {
       // We have found that the target thread is inside of a scoped access.
       // An asynchronous handshake is sent to the target thread, telling it
       // to throw an exception, which will unwind the target thread out from
       // the scoped access.
       OopHandle session(Universe::vm_global(), JNIHandles::resolve(_session));
       OopHandle error(Universe::vm_global(), JNIHandles::resolve(_error));
       jt->install_async_exception(new ScopedAsyncExceptionHandshake(session, error));
+    } else if (!in_scoped) {
+      frame last_frame = get_last_frame(jt);
+      if (last_frame.is_compiled_frame() && last_frame.can_be_deoptimized()) {
+        // We are not at a safepoint that is 'in' an @Scoped method, but due to the compiler
+        // moving code around/hoisting checks, we may be in a situation like this:
+        //
+        // liveness check (from @Scoped method)
+        // for (...) {
+        //    for (...) { // strip-mining inner loop
+        //        memory access (from @Scoped method)
+        //    }
+        //    safepoint <-- STOPPED HERE
+        // }
+        //
+        // The safepoint at which we're stopped may be in between the liveness check
+        // and actual memory access, but is itself 'outside' of @Scoped code
```

Comment:
what is `@Scoped code`? I don't see that annotation mentioned here: https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/ScopedValue.html

Reply:
This is the whole magic around the shared arena. It is not public API and internal to Hotspot/VM:
- https://github.com/openjdk/jdk/blob/a96de6d8d273d75a6500e10ed06faab9955f893b/src/java.base/share/classes/jdk/internal/misc/X-ScopedMemoryAccess.java.template#L117-L119
- https://github.com/openjdk/jdk/blob/a96de6d8d273d75a6500e10ed06faab9955f893b/src/hotspot/share/prims/scopedMemoryAccess.cpp#L143-L149

Reply:
> what is `@Scoped code`? I don't see that annotation mentioned here: https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/ScopedValue.html

This is nothing to do with scoped values, instead this is an annotation declared in jdk.internal.misc.ScopedMemoryAccess that is known to the VM.

Reply:
Basically if the VM is inside a @Scoped method and it starts a thread-local handshake, it will deoptimize top-most frame of all those threads so they can do the "isAlive" check.